src/xdp-utils.c

```diff
@@ -144,11 +145,72 @@ xdp_app_info_new (XdpAppInfoKind kind)
   return app_info;
 }
 
+void
+decode_hex_escapes (char *str[])
+{
+  char hex[] = "0xhh";
+
+  for (int i=0; (*str)[i]!='\0' && (*str)[i+1]!='\0'; ++i)
+  {
+    if ((*str)[i] == '\\' && (*str)[i+1] == 'x')
+      {
+        hex[2] = (*str)[i+2];
+        hex[3] = (*str)[i+3];
+        char unescaped = (char) strtol (hex, NULL, 16);
+        (*str)[i+2] = '\\';
+        (*str)[i+3] = unescaped;
+        i += 3;
+      }
+  }
+  char **splits = g_strsplit(*str, "\\x\\", -1);
+  g_free (*str);
+  *str = g_strjoinv ("", splits);
+  g_strfreev (splits);
+  return;
+}
+
+void
+get_app_id_from_pid (pid_t pid, XdpAppInfo *app_info)
+{
+  char *unit, **dash_splits, **at_splits, **app_id;
+  int res, len;
+
+  res = sd_pid_get_user_unit(pid, &unit);
+  /*
+   * the session might not be managed by systemd
+   * or there could be error fetching own systemd units
+   * or the unit might not be started by the the desktop environment (e.g. it's a script run from terminal)
+   */
+  if (res == -ENODATA || res < 0 || !g_str_has_prefix (unit, "app-"))
+    {
+      app_info->id = g_strdup ("");
+      return;
+    }
+  dash_splits = g_strsplit (unit, "-", -1);
+  len = g_strv_length (dash_splits);
+  /*
+   * format:
+   * app[-<launcher>]-<ApplicationID>-<RANDOM>.scope
+   * app[-<launcher>]-<ApplicationID>-autostart.service -> no longer true since systemd v248
```

Comment:
I think instead of relying on the part after the app-ID to not contain any additional `-` (e.g. `app-gnome-com.example.Foobar-123-456.service` would break this logic), it might make more sense to assume that the `<launcher>` does not contain dots (in practice it's something like `gnome`, `flatpak` or `steam`, which doesn't), so that the logic could be:

* remove trailing `.service`, `.slice` or `.scope`
* remove an `@` and anything after it, if present
* split into dash-separated components
* if second component contains at least two dots, it's the app ID
* else if third component contains at least two dots, it's the app ID

Reply:
I came up with a regex that seems to work for this: `"^app-(?:[[:alnum:]]+\\-)?(.+?\\..+?\\..+?)(?:[@\\-][0-9]*)?(?:-autostart)?(?:\\.scope|\\.service|\\.slice)$"`

But unfortunately the unit returned by systemd for gnome-software (which runs as `/usr/bin/gnome-software --gapplication-service`) is `app-gnome-gnome\x2dsoftware\x2dservice-9502.scope` (where 9502 is the pid in my case) so parsing that in any way is not going to get us the correct app ID of `org.gnome.Software`.